testsuite/integration/manualmode/src/test/java/org/jboss/as/test/manualmode/undertow/CustomUndertowFilterTestCase.java

```diff
@@ -0,0 +1,206 @@
+        } finally {
+            CONTROLLER.stop();
+        }
+    }
+
+    private static void prepareServerConfiguration() throws Exception {
+        try (final CLIWrapper cli = new CLIWrapper(true)) {
+            // create a (JBoss) module jar containing the custom http handler and a dummy class that gets used
+            // by the handler
+            final Path moduleJar = createJar("custom-http-handler", CustomHttpHandler.class, SomeClassInSameModuleAsCustomHttpHandler.class);
+            try {
+                cli.sendLine("module add --name=" + CUSTOM_FILTER_MODULE_NAME
+                        + " --slot=main --dependencies=io.undertow.core --resources="
+                        + moduleJar.toAbsolutePath());
+            } finally {
+                Files.deleteIfExists(moduleJar);
+            }
+            // create a custom-filter in the undertow subsystem and use the filter class that's
+            // present in the module that we just created
+            cli.sendLine(String.format(
+                    "/subsystem=undertow/configuration=filter/custom-filter=%s:add(class-name=%s, module=%s)",
+                    CUSTOM_FILTER_RESOURCE_NAME, CUSTOM_FILTER_CLASSNAME, CUSTOM_FILTER_MODULE_NAME));
+            // add a reference to this custom filter, in the default undertow host, so that it gets used
+            // for all deployed applications on this host
+            cli.sendLine(String.format(
+                    "/subsystem=undertow/server=default-server/host=default-host/filter-ref=%s:add()",
+                    CUSTOM_FILTER_RESOURCE_NAME));
+            // reload the server for changes to take effect
+            ServerReload.executeReloadAndWaitForCompletion(CONTROLLER.getClient().getControllerClient());
+        }
+    }
+
+    private static void resetServerConfiguration() throws Exception {
+        try (final CLIWrapper cli = new CLIWrapper(true)) {
+            // remove the filter-ref
+            cli.sendLine(String.format(
+                    "/subsystem=undertow/server=default-server/host=default-host/filter-ref=%s:remove()",
+                    CUSTOM_FILTER_RESOURCE_NAME));
+            // remove the custom undertow filter
+            cli.sendLine(String.format(
+                    "/subsystem=undertow/configuration=filter/custom-filter=%s:remove()",
+                    CUSTOM_FILTER_RESOURCE_NAME));
+            // remove the module containing the filter
+            cli.sendLine("module remove --name=" + CUSTOM_FILTER_MODULE_NAME);
+            // reload the server
+            ServerReload.executeReloadAndWaitForCompletion(CONTROLLER.getClient().getControllerClient());
+        }
+    }
+
+    private static void deploy() throws Exception {
+        // create a deployment
+        final Path warFilePath = createDeployment();
+        // deploy it to the server
+        try (final CLIWrapper cli = new CLIWrapper(true)) {
+            cli.sendLine(String.format(
```

Comment:
for deployment, it is better to use ServerDeploymentHelper to which you pass ManagementClient
and than directly call #deploy on it.